oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentPropertyStateTest.java

```diff
@@ -81,4 +114,289 @@ public void multiValuedBinarySize() throws Exception {
+        getBrokenSurrogateAndInitializeDifferentStores(MONGO, false);
+    }
+
+    @Test
+    public void testBrokenSurrogateWithoutCompressionForRDB() throws CommitFailedException {
+        getBrokenSurrogateAndInitializeDifferentStores(RDB_H2, false);
+    }
+
+    @Test
+    public void testBrokenSurrogateWithoutCompressionForInMemory() throws CommitFailedException {
+        getBrokenSurrogateAndInitializeDifferentStores(MEMORY, false);
+    }
+
+    @Test
+    public void testBrokenSurrogateWithCompressionForMongo() throws CommitFailedException {
+        DocumentPropertyState.setCompressionThreshold(1);
+        getBrokenSurrogateAndInitializeDifferentStores(MONGO, true);
+    }
+
+    @Test
+    public void testBrokenSurrogateWithCompressionForRDB() throws CommitFailedException {
+        DocumentPropertyState.setCompressionThreshold(1);
+        getBrokenSurrogateAndInitializeDifferentStores(RDB_H2, true);
+    }
+
+    @Test
+    public void testBrokenSurrogateWithCompressionForInMemory() throws CommitFailedException {
+        DocumentPropertyState.setCompressionThreshold(1);
+        getBrokenSurrogateAndInitializeDifferentStores(MEMORY, true);
+    }
+
+    private void getBrokenSurrogateAndInitializeDifferentStores(DocumentStoreFixture fixture, boolean compressionEnabled) throws CommitFailedException {
+        String test = "brokensurrogate:dfsa\ud800";
+        DocumentStore store = null;
+        DocumentNodeStore nodeStore = null;
+
+        try {
+            store = fixture.createDocumentStore();
+
+            if (store instanceof MongoDocumentStore) {
+                // Enforce primary read preference, otherwise tests may fail on a
+                // replica set with a read preference configured to secondary.
+                // Revision GC usually runs with a modified range way in the past,
+                // which means changes made it to the secondary, but not in this
+                // test using a virtual clock
+                MongoTestUtils.setReadPreference(store, ReadPreference.primary());
+            }
+            nodeStore = new DocumentMK.Builder().setDocumentStore(store).getNodeStore();
+
+            createPropAndCheckValue(nodeStore, test, compressionEnabled);
+
+        } finally {
+            if (nodeStore != null) {
+                nodeStore.dispose();
+            }
+            if (store != null) {
+                store.dispose();
+            }
+        }
+
+    }
+
+    private void createPropAndCheckValue(DocumentNodeStore nodeStore, String test, boolean compressionEnabled) throws CommitFailedException {
+        NodeBuilder builder = nodeStore.getRoot().builder();
+        if (compressionEnabled) {
+            DocumentPropertyState.setCompressionThreshold(1);
+        }
+        DocumentPropertyState documentPropertyState = new DocumentPropertyState(nodeStore, "p", test, Compression.GZIP);
+        builder.child(TEST_NODE).setProperty("p", documentPropertyState.getValue(), Type.STRING);
```

Comment:
```suggestion
        builder.child(TEST_NODE).setProperty("p", test, Type.STRING);
```

As mentioned in the [previous comment](https://github.com/apache/jackrabbit-oak/pull/1526#issuecomment-2236322219), going via DocumentPropertyState when setting a property on a NodeBuilder is not correct (I might have made that point not fully clear there).

With this fix, the tests actually succeed

Reply:
Done. Thanks.